src/carnot/funcs/builtins/pprof_ops.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <absl/container/flat_hash_map.h>
+
+#include <string>
+
+#include "src/carnot/udf/registry.h"
+#include "src/carnot/udf/udf.h"
+#include "src/shared/pprof/pprof.h"
+
+namespace px {
+namespace carnot {
+namespace builtins {
+
+using px::shared::PProfProfile;
+
+class CreatePProfRowAggregate : public udf::UDA {
+ public:
+  static udf::UDADocBuilder Doc() {
+    return udf::UDADocBuilder("Convert perf profiling data to pprof format.")
+        .Details("Converts perf profiling stack traces into pprof format.")
+        .Example(
+            R"example(stack_traces = px.DataFrame(table='stack_traces.beta', start_time='-1m')
+sample_period = px.GetProfilerSamplingPeriodMS()
+stack_traces.asid = px.asid()
+df = stack_traces.merge(sample_period, how='inner', left_on=['asid'], right_on=['asid'])df = df.groupby(['profiler_sampling_period_ms'])
+df = df.agg(pprof=('stack_trace', 'count', 'profiler_sampling_period_ms', px.pprof)))example")
+        .Arg("stack_trace", "Stack trace string.")
+        .Arg("count", "Count of the stack trace string.")
+        .Arg("profiler_period_ms", "Profiler stack trace sampling period in ms.")
+        .Returns("A single row that aggregates all the stack traces and counts into pprof format.");
+  }
+
+  void Update(FunctionContext*, const StringValue stack_trace, const Int64Value count,
+              const Int64Value profiler_period_ms) {
+    UpdateOrCheckSamplingPeriod(profiler_period_ms.val);
+
+    histo_[stack_trace] += count.val;
+  }
+
+  void Merge(FunctionContext*, const CreatePProfRowAggregate& other) {
+    UpdateOrCheckSamplingPeriod(other.profiler_period_ms_);
+
+    for (const auto& [stack_trace, count] : other.histo_) {
+      histo_[stack_trace] += count;
+    }
+  }
+
+  StringValue Serialize(FunctionContext*) {
+    if (multiple_profiler_periods_found_) {
+      return "Protobuf `SerializeToString` failed, multiple profiling periods found.";
+    }
+
+    const auto pprof = px::shared::CreatePProfProfile(profiler_period_ms_, histo_);
+    std::string output;
+    const bool ok = pprof.SerializeToString(&output);
+    if (!ok) {
+      return "Protobuf `SerializeToString` failed.";
+    }
+    return output;
+  }
+
+  Status Deserialize(FunctionContext*, const StringValue& upstream_pprof_str) {
```

Comment:
minor nit: I don't like the name `upstream_pprof_str`. `upstream` is kinda confusing.
rename to `pprof_str`